### PR TITLE
3d advection

### PR DESCRIPTION
**Pull request by colinjcotter**
Add code to do velocity advection in 3D.

Also, automatically build broken spaces for Embedded DG.

### CODE REVIEW

**Comment by jshipton** [comment]
Does this need firedrake to be updated to pass tests? How do we do that on travis?

**Comment by colinjcotter** [comment]
Oh yes, that might be what's needed.

On 13 July 2016 at 13:28, jshipton <notifications@github.com> wrote:

> Does this need firedrake to be updated to pass tests? How do we do that on
> travis?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/gusto/pull/51#issuecomment-232340177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF0q8spzPHRA-wFm7mcOgPCq1BURl4BDks5qVNoFgaJpZM4JIHzs>
> .
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916

**Comment by wence-** [comment]
No, these are actual bugs.  To make a VectorFunctionSpace you need the `VectorElement` modifier on the outside of the element.  To make the `BrokenElement` you need to unwrap the passed in element and wrap it up again, I think.

File: gusto/advection.py [hunk 1]
```diff
@@ -191,18 +192,30 @@ def apply(self, x_in, x_out):
 
 class EmbeddedDGAdvection(DGAdvection):
 
-    def __init__(self, state, Vdg, continuity):
+    def __init__(self, state, V, Vdg=None, continuity=False):
```
Comment (wence-) [inline gusto/advection.py]: You've changed the constructor signature, but haven't updated the embedded dg advection tests, which still use the old signature.

**Comment by wence-** [comment]
I think I fixed this.  Note that firedrake is always rebuilt for gusto on travis, so you don't have to worry about firedrake being out of date.

**Comment by jshipton** [comment]
Don't you need the function space, rather than the function? And anyway, I still get the same error.

**Comment by wence-** [comment]
Oh, maybe.  I don't know what the return values of all these things are.

**Comment by wence-** [comment]
I think I've now fixed the remaining embeddedDG constructor calls.

**Comment by wence-** [comment]
Green and merged.

**Comment by jshipton** [comment]
Hang on a minute! The new code isn't actually tested... I'm just sorting that out now.

________________________________

**Comment by wence-** [comment]
Too late :p.  Just make a new branch adding tests.

**Comment by jshipton** [comment]
:P


I'm getting muddled using xfail with parameterise. At the moment I have:


@pytest.mark.parametrize("vector", [False, True])

@pytest.mark.parametrize("continuity", [False, True])

@pytest.mark.parametrize("broken", [False, True])


and I want to mark the ones with "broken" = True and "vector" = True as xfails.


I've tried amending the last parameterise to:


@pytest.mark.parametrize("broken", [False, pytest.mark.xfail("vector" is True, reason="not implemented")(True)])


but it seems to be ignoring the xfail mark - maybe it doesn't know what "vector" is yet?


Any ideas?


J

________________________________

**Comment by wence-** [comment]
There's no way to access the parameter value inside the parametrize decorator.  Here might be one way:

```python
import pytest

params = [False, True]

@pytest.fixture(params=params)
def vector(request):
    return request.param

@pytest.fixture(params=params)
def continuity(request):
    return request.param

@pytest.fixture(params=params)
def broken(request, vector):
    if request.param and vector:
        request.applymarker(pytest.mark.xfail(reason="foo"))
    return request.param

def test_foo(vector, continuity, broken):
    if vector and broken:
        assert False
    else:
        assert True
```

**Pull request merged**